File: app/src/main/java/com/gameautopilot/app/vision/SetOfMarksOverlay.kt
package com.gameautopilot.app.vision

import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Rect
import com.gameautopilot.app.core.MarkBox
import kotlin.math.max
import kotlin.math.min

/**
 * Draws translucent numbered rectangles on top of a bitmap copy.
 * The numbers correspond to MarkBox.id; the brain returns these IDs
 * via {"type":"tapMark","markId":N} instead of raw pixel coordinates.
 */
object SetOfMarksOverlay {

    fun annotate(src: Bitmap, marks: List<MarkBox>): Bitmap {
        val out = src.copy(Bitmap.Config.ARGB_8888, true)
        if (marks.isEmpty()) return out
        val canvas = Canvas(out)
        val box = Paint(Paint.ANTI_ALIAS_FLAG).apply {
            style = Paint.Style.STROKE
            strokeWidth = max(2f, src.width / 480f)
            color = Color.argb(220, 80, 220, 120)
        }
        val labelBg = Paint(Paint.ANTI_ALIAS_FLAG).apply {
            style = Paint.Style.FILL
            color = Color.argb(220, 0, 0, 0)
        }
        val labelText = Paint(Paint.ANTI_ALIAS_FLAG).apply {
            color = Color.argb(255, 255, 255, 120)
            textSize = max(18f, src.width / 50f)
            isFakeBoldText = true
        }
        val pad = max(4f, src.width / 240f)

        for (m in marks) {
            canvas.drawRect(m.left.toFloat(), m.top.toFloat(), m.right.toFloat(), m.bottom.toFloat(), box)
            val text = m.id.toString()
            val bounds = Rect()
            labelText.getTextBounds(text, 0, text.length, bounds)
            val bgLeft = m.left.toFloat()
            val bgTop = max(0f, m.top.toFloat() - bounds.height() - 2f * pad)
            val bgRight = bgLeft + bounds.width() + 2f * pad
            val bgBottom = bgTop + bounds.height() + 2f * pad
            canvas.drawRect(bgLeft, bgTop, bgRight, bgBottom, labelBg)
            canvas.drawText(text, bgLeft + pad, bgBottom - pad - 2f, labelText)
        }
        return out
    }
}

@Suppress("unused")
private fun clampi(v: Int, lo: Int, hi: Int) = min(max(v, lo), hi)

File: mac/Sources/GameAutopilotMac/Perception/SetOfMarksOverlay.swift
import AppKit
import CoreGraphics
import Foundation

/// Draws translucent numbered rectangles on top of a copy of the captured
/// image. The numbers correspond to MarkBox.id; the brain returns these
/// ids via {"type":"tapMark","markId":N} instead of raw pixel coordinates.
/// Mirrors vision/SetOfMarksOverlay.kt (Android Canvas -> CoreGraphics).
enum SetOfMarksOverlay {
    static func annotate(_ image: CGImage, marks: [MarkBox]) -> CGImage {
        guard !marks.isEmpty else { return image }

        let width = image.width
        let height = image.height
        guard let context = CGContext(
            data: nil,
            width: width,
            height: height,
            bitsPerComponent: 8,
            bytesPerRow: 0,
            space: CGColorSpaceCreateDeviceRGB(),
            bitmapInfo: CGImageAlphaInfo.premultipliedLast.rawValue
        ) else { return image }

        // Deliberately NOT flipping the CTM here: CGContext.draw(_:in:)
        // already orients a CGImage correctly in a context's native
        // (bottom-left origin, Y-up) space -- adding our own flip on top
        // would turn the base screenshot upside down. MarkBox coordinates
        // are top-left-origin/Y-down (matching AX/CGEvent/screenshot pixel
        // space), so each shape below converts its own Y explicitly
        // (contextY = imageHeight - pixelSpaceY) instead of touching the CTM.
        context.draw(image, in: CGRect(x: 0, y: 0, width: width, height: height))

        let strokeWidth = max(2.0, Double(width) / 480.0)
        let fontSize = max(14.0, Double(width) / 60.0)
        let pad = max(3.0, Double(width) / 300.0)
        let h = Double(height)

        context.setLineWidth(strokeWidth)
        context.setStrokeColor(CGColor(red: 0.31, green: 0.86, blue: 0.47, alpha: 0.9))
        for m in marks {
            let rect = CGRect(x: Double(m.left), y: h - Double(m.bottom),
                               width: Double(m.width), height: Double(m.height))
            context.stroke(rect)
        }

        // This context is genuinely un-flipped (native Quartz Y-up), so
        // tell AppKit's text system exactly that.
        let nsContext = NSGraphicsContext(cgContext: context, flipped: false)
        NSGraphicsContext.saveGraphicsState()
        NSGraphicsContext.current = nsContext

        let font = NSFont.boldSystemFont(ofSize: fontSize)
        let attrs: [NSAttributedString.Key: Any] = [.font: font, .foregroundColor: NSColor.white]

        for m in marks {
            let attributed = NSAttributedString(string: "\(m.id)", attributes: attrs)
            let textSize = attributed.size()
            let labelHeight = textSize.height + pad * 2

            // Label sits just above the mark's top edge, in image space --
            // i.e. its context-space Y starts at the mark's converted top
            // and extends further upward (larger Y) by its own height.
            // Clamp so a mark near the top of the frame doesn't push the
            // label off-canvas.
            let labelBottomInContext = min(h - Double(m.top), h - labelHeight)
            let bgRect = CGRect(
                x: Double(m.left),
                y: labelBottomInContext,
                width: textSize.width + pad * 2,
                height: labelHeight
            )
            NSColor.black.withAlphaComponent(0.78).setFill()
            bgRect.fill()
            attributed.draw(at: CGPoint(x: bgRect.origin.x + pad, y: bgRect.origin.y + pad))
        }

        NSGraphicsContext.restoreGraphicsState()

        return context.makeImage() ?? image
    }
}
